cpp/include/mun/reflection.h

```diff
@@ -10,74 +10,43 @@
 #include "mun/type_info.h"
 
 namespace mun {
-constexpr inline bool operator==(const MunTypeId& lhs, const MunTypeId& rhs) noexcept {
-    for (auto idx = 0; idx < 16; ++idx) {
-        if (lhs.guid._0[idx] != rhs.guid._0[idx]) {
-            return false;
-        }
-    }
-    return true;
-}
-
-constexpr inline bool operator!=(const MunTypeId& lhs, const MunTypeId& rhs) noexcept {
-    return !(lhs == rhs);
-}
 
 template <typename T>
 struct ArgumentReflection {
-    static constexpr const char* type_name(const T&) noexcept { return StaticTypeInfo<T>::name(); }
-    static constexpr MunTypeId type_id(const T&) noexcept { return StaticTypeInfo<T>::id(); }
+    static TypeInfo type_info(const T&) noexcept { return StaticTypeInfo<T>::type_info(); }
 };
 
 template <typename T>
 struct ReturnTypeReflection {
-    static constexpr const char* type_name() noexcept { return StaticTypeInfo<T>::name(); }
-    static constexpr MunTypeId type_id() noexcept { return StaticTypeInfo<T>::id(); }
+    static bool accepts_type(const TypeInfo& ty) noexcept {
+        return StaticTypeInfo<T>::type_info() == ty;
+    }
+    static std::string_view type_hint() { return StaticTypeInfo<T>::type_info().name(); }
 };
 
 namespace reflection {
-template <typename T, typename U>
-constexpr bool equal_types() noexcept {
-    return ReturnTypeReflection<T>::type_id() == ReturnTypeReflection<U>::type_id();
-}
-
 template <typename Arg>
-inline std::optional<std::pair<const char*, const char*>> equals_argument_type(
+inline std::optional<std::pair<std::string, std::string>> equals_argument_type(
     const TypeInfo& type_info, const Arg& arg) noexcept {
-    if (type_info.id() == ArgumentReflection<Arg>::type_id(arg)) {
+    auto arg_type_info = ArgumentReflection<Arg>::type_info(arg);
+    if (arg_type_info == type_info) {
         return std::nullopt;
     } else {
-        const auto expected_name = ArgumentReflection<Arg>::type_name(arg);
-        return std::make_pair(type_info.name().data(), expected_name);
+        const auto expected_name = arg_type_info.name();
+        return std::make_pair(std::string(type_info.name()), std::string(expected_name));
     }
 }
-
 template <typename T>
-inline std::optional<std::pair<const char*, const char*>> equals_return_type(
-    const TypeInfo& type_info) noexcept;
-
-}  // namespace reflection
-
-}  // namespace mun
-
-#include "struct_ref.h"
-
-namespace mun {
-namespace reflection {
-template <typename T>
-std::optional<std::pair<const char*, const char*>> equals_return_type(
+inline std::optional<std::pair<std::string, std::string>> equals_return_type(
```

Comment:
Would string_view not be cheaper?

Reply:
Yes, but that also doesn't specify anything about the lifetime of the contents. This is a generic function with no guarantee about the lifetime of the string data. Therefore I changed this to use a string.